dask_expr/_core.py

```diff
@@ -20,13 +20,18 @@ class Expr:
     _defaults = {}
 
     def __init__(self, *args, **kwargs):
+        from dask_expr._collection import FrameBase, Scalar
+
         operands = list(args)
         for parameter in type(self)._parameters[len(operands) :]:
             try:
                 operands.append(kwargs.pop(parameter))
             except KeyError:
                 operands.append(type(self)._defaults[parameter])
         assert not kwargs, kwargs
+        operands = [
+            o.expr if isinstance(o, (FrameBase, Scalar)) else o for o in operands
+        ]
```

Comment:
Maybe instead we check for some `__dask_expr__` attribute?  This seems like a case where a protocol might be better than type checking.

Reply:
I am Not a big fan of these type of protocols, that said the scheduler integration will make this specific issue go away so that’s only a temporary solution anyway

Reply:
I think that protocols are more aligned with what you actually want here.  You want to exprify anything that knows how to turn itself into an expr.  You don't actually care about the FrameBase or Scalar classes.  

If you're against protocols in principle in some way then we should probably talk about that.  A lot of Dask depends on protocols.  Going against the grain there is likely to cause friction down the road.

Reply:
I am not against protocols in principle but I am cautious about them. They have caused me pain in the past through thinks like what we have with __dask_graph__ where a too expensive computation might get triggered if you are not careful or through catching things that you don’t actually want to catch.

Checking for an expr attribute here should be fine though

Reply:
> They have caused me pain in the past through thinks like what we have with dask_graph where a too expensive computation might get triggered if you are not careful or through catching things that you don’t actually want to catch

Yeah, in that case I think it has a lot more to do with how we implemented things than with the protocol itself.  Certainly I'm not in favor of doing things badly :)